WIP: "modernize" single-source versioning (#1273)
pypa/packaging.python.org: source/guides/single-sourcing-package-version.rst

===================================

There are many techniques to maintain a single source of truth for the version string of your projects:

Auto extract the version string from the source
-----------------------------------------------

One way of single sourcing is to specify the version string somewhere in the source code, e.g. ``my_package/__init__.py``. Then it can be found at runtime with ``my_package.__version__``, and that same value can be used to set the version when building the package.

> #. Declare to read the version string  from the source in ``setuptools.cfg`` or ``pyproject.toml``

Reviewer: Typo
```suggestion
#. Declare to read the version string  from the source in :file:`setup.cfg` or `pyproject.toml`.
```

Reviewer 2: The plan from setuptools side is to [eventually deprecate `setup.cfg`](https://github.com/pypa/setuptools/issues/3214), so it should not be mentioned. There's also spurious spacing here between "string  from".

Reviewer 3: > The plan from setuptools side is to [eventually deprecate `setup.cfg`](https://github.com/pypa/setuptools/issues/3214),

Oh? Really? That's sad. I do not want to have a `pyproject.toml` file that is kilometers long. I like having one configuration file per tool. It is easier to find things. I just wished there was maybe a convention like `pyproject.d/setuptools.toml`.

Reviewer 4: > Maybe it would be good to have a convention like `pyproject.d/setuptools.toml`.

That sounds like a great standard to have!

However it'd require a huge push to create and support it by the various packaging tools. Could you make a proposal over at the `peps` repository?

   With recent versions of setuptools (since 46.4.0), one can add a declaration
   to the project's :file:`setup.cfg` file (replacing "package" with the import
   name of the package):

   .. code-block:: ini

       [metadata]
       version = attr: package.__version__

   As of the release of setuptools 61.0.0, one can also specify the
   version dynamically in the project's :file:`pyproject.toml` file.

   .. code-block:: toml

        [project]
        name = "package"
        dynamic = ["version"]

        [tool.setuptools.dynamic]
        version = {attr = "package.__version__"}

   Please be aware that declarative config indicators, including the
   ``attr:`` directive, are not supported in parameters to
   :file:`setup.py`.

#.  While the modern standard is to use declarative files (such as ``setup.cfg``),
    with older versions of setuptools, you may need to add a version-reading

>     function to setup.py: Example adapted from (from `pip setup.py <https://github.com/pypa/pip/blob/main/setup.py#L11>`_)::

Reviewer: Let's pin the URL since it may become 404 if pip ever removes this file from its main branch.
```suggestion
    function to :file`setup.py`: Example adapted from (from `pip setup.py <https://github.com/pypa/pip/blob/6f3a7181b6bfa0e95d479b3269baa0e551ff2cb2/setup.py#L11>`_)::
```

        from pathlib import Path

        def read(rel_path):

>             here = Path(__file__).parent.absolute()

Reviewer 2: Why calling absolute method? `__file__` is an absolute path already since [Python 3.4](https://docs.python.org/3.4/whatsnew/3.4.html#other-language-changes).

Reviewer: Many confuse `absolute()` and `resolve()`: https://discuss.python.org/t/pathlib-absolute-vs-resolve/2573/2.
We should use the latter:
```suggestion
            here = Path(__file__).parent.resolve()
```

@wimglenn to answer your question, `__file__` is not what's being unwound on this line — the parent directory is. Also, absolute and resolved are different concepts. Absolute just makes the path start with `/` so it's not relative. The resolution process follows symlinks, revealing actual location in the file system.

That said, I don't know if anything would change, realistically, if we were to skip the resolution...

Reviewer 2: Can you show me an example where resolving the parent of `Path(__file__)` is necessary?

Reviewer: I would imagine that probably not in the context of Python packaging. In general, it's probably useful when paths are processed by different programs with different access privileges.

Author: I've changed it to `.resolve` -- why not?

Reviewer 2: Why not: because it should be no more complicated than necessary. I'm not convinced resolving the path is ever necessary.

>             with open(here / rel_path), 'r') as fp:

Reviewer: Since this is already injecting pathlib, just use the high-level API instead:
```suggestion
            return (here / rel_path).read_text(encoding='utf-8')
```

> ===================================

Reviewer 5: ```suggestion
===================================
```

Auto extract the version string from the source
-----------------------------------------------

One way of single sourcing is to specify the version string somewhere in the source code, e.g. ``my_package/__init__.py``. Then it can be found at runtime with ``my_package.__version__``, and that same value can be used to set the version when building the package.

#. Declare to read the version string from the source in ``pyproject.toml`` or ``setuptools.cfg``

   As of the release of setuptools 61.0.0, one can specify the
   version dynamically in the project's :file:`pyproject.toml` file.

   .. code-block:: toml

        [project]
        name = "my_package"
        dynamic = ["version"]

        [tool.setuptools.dynamic]
        version = {attr = "my_package.__version__"}

   Please be aware that declarative config indicators, including the
   ``attr:`` directive, are not supported in parameters to :file:`setup.py`.

   The declaration can alternatively be added to the project's :file:`setup.cfg` file (replacing "my_package" with the import name of the package):

   .. code-block:: ini

       [metadata]
       version = attr: my_package.__version__

>    Note that the use of :file:`setup.cfg` is being deprecated by setuptools -- new projects should use :file:`pyproject.toml`.

Reviewer 2: Where was that established? As far as I know the plan from setuptools side is still to "eventually deprecate" setup.cfg, because it does not make sense to have two different declarative formats for the same purpose.

Ref: https://github.com/pypa/setuptools/issues/3214

>    Note that the use of :file:`setup.cfg` is being deprecated by setuptools -- new projects should use :file:`pyproject.toml`.

Reviewer 6: Perhaps a good way to handle this is to put a note block in here and avoid deprecation comment here since it is not necessary for the reader.

```
In general, :file:`pyproject.toml` should be used for new projects.
```

>    Note that the use of :file:`setup.cfg` is being deprecated by setuptools -- new projects should use :file:`pyproject.toml`.

Reviewer 7: I don’t understand your proposed sentence. “stick with” means to keep using an old thing, no? `pyproject.toml` is the new thing, so you can’t “stick with” it.

>    Note that the use of :file:`setup.cfg` is being deprecated by setuptools -- new projects should use :file:`pyproject.toml`.

Reviewer: You're right. Perhaps, I had a slightly different context in my head, thinking of a situation when there's a `pyproject.toml`, I want to separate it into a dedicated config (which would have to be `setup.cfg`) and somebody would want to keep using `pyproject.toml`.

Let's try rephrasing:
```suggestion
   Note that many people prefer :file:`pyproject.toml` over :file:`setup.cfg`, while some still choose to keep using the latter regardless.
```

===================================

There are many techniques to maintain a single source of truth for the version string of your projects:

Auto extract the version string from the source
-----------------------------------------------

One way of single sourcing is to specify the version string somewhere in the source code, e.g. ``my_package/__init__.py``. Then it can be found at runtime with ``my_package.__version__``, and that same value can be used to set the version when building the package.

#. Declare to read the version string from the source in ``pyproject.toml`` or ``setuptools.cfg``

>    As of the release of setuptools 61.0.0, one can also specify the

Reviewer 3: I would remove the "also" here.

   version dynamically in the project's :file:`pyproject.toml` file.

   .. code-block:: toml

        [project]
        name = "package"
        dynamic = ["version"]

        [tool.setuptools.dynamic]

>         version = {attr = "package.__version__"}

Reviewer 3: Should be `my_package` instead of `package` since that is the example used earlier in the page.

   Please be aware that declarative config indicators, including the
   ``attr:`` directive, are not supported in parameters to
   :file:`setup.py`.

>    With older versions of setuptools the declaration can be added to

Reviewer 3: Why the "with older versions of setuptools"? Current versions support this perfectly well as far as I know.

Author: The idea was that you should use pyproject.toml with new setuptools -- but yes, that is confusing -- and there is a lot of setup.cfg out there still. I"ll rephrase.

>    the project's :file:`setup.cfg` file (replacing "package" with the import

Reviewer 3: `my_package`

   name of the package):

   .. code-block:: ini

       [metadata]

>        version = attr: package.__version__

Reviewer 3: `my_package`

>     Note that the use of :file:`setup.cfg` is being depricated by setuptools -- new projects should use :file:`pyproject.toml`.

Reviewer 4: This line is somehow swallowed by the INI code block.

Reviewer 3: Reference to the deprecation?

Also typo: *deprecated* instead of *depricated*.

Author: thanks -- fixed

#.  While the modern standard is to use declarative files (such as ``pyproject.toml``),

>     with older versions of setuptools, you may need to add a version-reading

Reviewer 3: *setuptools* will support `setup.py` forever. Maybe `setup.cfg` will disappear (hopefully not), but I am quite sure `setup.py` will stay for a very very long time.

Author: well, yes the the declarative forms are still preferred -- and you can use setup.cfg or pyproject.toml alongside setup.py, yes?

Author: > Reference to the deprecation?

Someone said that in this PR -- if that's not a done deal, we should remove it.

Reviewer: > _setuptools_ will support `setup.py` forever. Maybe `setup.cfg` will disappear (hopefully not), but I am quite sure `setup.py` will stay for a very very long time.

Agreed. It'd be unfortunate and these were deprecated fully. There are cases when one would want to keep metadata in a dedicated file. In case of mass-maintenance of many projects, for example.

There are many techniques to maintain a single source of truth for the version

> number of your projects:

Reviewer 3: I recommend using "*version string*" consistently (throughout this document, the whole `packaging.python.org` and in general).

Auto extract the version string from the source
-----------------------------------------------

> One way of single sourcing is to specify the version somewhere in the source code, e.g. ``my_package/__init__.py``. Then it can be found at runtime with ``my_package.__version__``, and that same value can be used to set the version when building the package.
> 
> #. Declare to read the version from the source in ``setuptools.cfg`` or ``pyproject.toml``
> 
>    With recent versions of setuptools (since 46.4.0), one can add a declaration
>    to the project's :file:`setup.cfg` file (replacing "package" with the import
>    name of the package):
> 
>    .. code-block:: ini
> 
>        [metadata]
>        version = attr: package.__version__
> 
>    As of the release of setuptools 61.0.0, one can also specify the
>    version dynamically in the project's :file:`pyproject.toml` file.
> 
>    .. code-block:: toml
> 
>         [project]
>         name = "package"
>         dynamic = ["version"]
> 
>         [tool.setuptools.dynamic]
>         version = {attr = "package.__version__"}
> 
>    Please be aware that declarative config indicators, including the
>    ``attr:`` directive, are not supported in parameters to
>    :file:`setup.py`.

Reviewer 8: How about we rename `__version__` to something less controversial. For example:

```suggestion
One way of single sourcing is to specify the version somewhere in the source code. Let's suppose you have a Python module (e.g. ``my_package/__init__.py``) that defines the project version as a constant (e.g. `VERSION = "0.42"`). Some build-backends allow you to use this value when building the package.

If you use `setuptools` as build-backend you can try one of the following:

#. Using the `version = attr:` in ``setup.cfg``:

   With recent versions of setuptools (since 46.4.0), one can add a declaration
   to the project's :file:`setup.cfg` file (replacing `my_package` with the import
   name of the package and `VERSION` with the name of the relevant attribute):

   .. code-block:: ini
       [metadata]
       version = attr: my_package.VERSION
       
#. Using a dynamic version in `pyproject.toml`:

   As of the release of setuptools 61.0.0, one can also specify the
   version dynamically in the project's :file:`pyproject.toml` file.

   .. code-block:: toml
        [project]
        name = "package"
        dynamic = ["version"]
        [tool.setuptools.dynamic]
        version = {attr = "package.__version__"}
        
Please be aware that declarative config indicators, including the
``attr:`` directive, are not supported in parameters to
:file:`setup.py`.
```

Reviewer 3: Do we really want to have examples specific to a build back-end? Isn't there a way to stay agnostic? At the very least we could use tabs like in the packaging tutorial.

Reviewer: > wouldn't it be nice if there were a standard way to specify how to set yourversion other than hard coding it in a toml file?

`attr:` is setuptools-specific, but `dynamic = ["version"]` allows all the PEP 517 build backends to generate the version during the build time. How this is done and how it's configured, it's always going to be specific to each build backend.
PEP 517 describes the communication interface between the parties but it doesn't mandate which backend-specific features should be implemented since it's out of the scope. What's the point of a standard that stays away from the implementation details if it'd demand each backend to implement the same mechanism? If that was the case, we wouldn't have the multitude of backends — just one.
Also, imagine a project being packaged is a wrapper around a Rust crate. And the version is in that crate. A packager uses a build backend specifically targeting Rust projects. If a backend is asked to implement loading the version from Python importables, it becomes impossible since that contradicts its design.

So the `dynamic` thing is really the extent of the standard that cares about packaging interoperability and doesn't have leaky abstractions.

Reviewer: > > How about we rename **version** to something less controversial.
> 
> `__version__` may be controversial, but it's also a very widely used convension -- seems a bit silly to not use the most common convention as the example.

I agree — to me, anything else would be controversial. I'd keep it as it is. But we can add a `.. tip::` admonition stating that "`__version__` is widely used and is sometimes considered controversial".

#.  While the modern standard is to use declarative files (such as ``setup.cfg``),
    with older versions of setuptools, you may need to add a version-reading
    function to setup.py: Example adapted from (from `pip setup.py <https://github.com/pypa/pip/blob/main/setup.py#L11>`_)::

        from pathlib import Path

        def read(rel_path):
            here = Path(__file__).parent.absolute()

>             with open(here / rel_path), 'r') as fp:

Reviewer 8: Not sure how much people need to rely on `setuptools<46.4.0` nowadays. Maybe we can omit this section?

Reviewer 8: Maybe we could replace this with the use case the dev needs to process a given file to extract the version string (e.g. the cases when you have a plain text file with several info or a file written in a different language?).

Reviewer: > Not sure how much people need to rely on `setuptools<46.4.0` nowadays. Maybe we can omit this section?

Pip-users? Not many. This'd mostly hit downstream packagers that might be stuck with old pinned packages in their distros.

.. NOTE2: Other back-end examples we should include?

Use an external version management tool
---------------------------------------

An external build tool can be used that manages the version string in both the SCM and source code, either directly or via an API. For example (not a complete list):

 - `commitizen <https://pypi.org/project/commitizen>`_,

 - `versioneer <https://github.com/python-versioneer/python-versioneer>`_,

 - `zest.releaser <https://pypi.org/project/zest.releaser>`_,

.. NOTE: are there others that should be included?

Dedicated file for the version string
-------------------------------------

Place the value in a simple :file:`VERSION` text file and have both
:file:`setup.py` and the project code read it.

::

    version = (mypackage_root_dir / 'VERSION').read_text(encoding='utf-8').strip()

An advantage with this technique is that it's not specific to Python.  Any
tool can read the version string.

.. warning::

    With this approach you must make sure that the ``VERSION`` file is included in
    all your source and binary distributions (e.g. add ``include VERSION`` to your
    :file:`MANIFEST.in`).

Nearly extinct methods
----------------------

> These methods rely on importing of code during the build process, or dynamically generating the version string at build time with custom code. These methods are prone to errors and security issues, but you may encounter them in older code bases.

Reviewer: Per https://github.com/pypa/packaging.python.org/pull/1273#discussion_r1407774382, I don't think that there's inherent security issue implications here. This statement is not universally true.

We're talking about the build scripts, and they run dynamic stuff by definition. Even when it's abstracted away into external libs. A more realistic attack surface is the supply chain — poisoning the deps is much more effective and simpler, then checking in things into a project since they go through code reviews much more often, then the deps updates. This is why, in the articles related to publishing from GHA to PyPI, for example, we always emphasize that building must happen in a separate CI job that wouldn't be able to elevate privileges and publish packages using legitimate credentials.

I think, this paragraph should instead focus on the UX/DX/maintainability/elegance aspects of the approach.

           ...
        )

Generate the version string from the SCM
----------------------------------------

Most projects use a Source Code Management System (SCM), such as git or mercurial to manage the code and also manage releases.
In order to keep the versioning of releases in sync with your package, the version string can be kept in the tags of an SCM, and automatically extracted with a tool such as
`setuptools_scm <https://pypi.org/project/setuptools-scm/>`_
or  `hatch-vcs <https://pypi.org/project/hatch-vcs/>`_

.. NOTE: maybe put in an example using setuptools_scm?

.. NOTE2: Other back-end examples we should include?

Use an external version management tool
---------------------------------------

An external build tool can be used that manages the version string in both the SCM and source code, either directly or via an API.

> A few tools you could use, in no particular order, and not necessarily complete:

Reviewer 3: the order is alphabetical, fair enough :)

and of course the list is not complete since we said "a few tools".

So I guess I would just leave it at "For example:"

Author: I borrowed that text from the original -- but I've simplified it -- but leaving the "not complete" in case anyone gets upset their favoratie tools wasn't included.

           ...
        )

Generate the version string from the SCM
----------------------------------------

> Most projects use a Source Code Management System (SCM), such as git or mercurial to manage the code and also manage releases.

Reviewer 2: VCS is a more familiar and less overloaded acronym than SCM

Reviewer: Honestly, I feel like it's vice versa. So this must be a subjective feeling.

Reviewer 2: Interesting, what is VCS overloaded for?  SCM is overloaded to [config management](https://en.wikipedia.org/wiki/Software_configuration_management).

Reviewer: I meant that it's more familiar.

Reviewer: Anyway, the overloading is comparable: https://en.wikipedia.org/wiki/VCS#Technology vs. https://en.wikipedia.org/wiki/SCM#Computing.

Author: I suppose VCS/SCM would work -- I used SCM because we later reference setuptools_scm.

Reviewer 2: But you also called it a version control system on line 73 :)

Author: fixed ...

> In order to keep the versioning of releases in sync with your package, the version string can be kept in the tags of a version control system, and automatically extracted with a tool such as

Reviewer: ```suggestion
In order to keep the versioning of releases in sync with your package, the version string can be kept in the tags of a version control system, and automatically extracted during the build, with a tool such as
```

`setuptools_scm <https://pypi.org/project/setuptools-scm/>`_.

.. NOTE: maybe put in an example using setuptools_scm?

> .. NOTE2: Is setuptools_scm the only one now?

Reviewer: For setuptools, yes. Other backends have their own plugins. Like hatch-vcs (which actually wraps setuptools-scm). There's also dunamai but it's a generic project, that doesn't provide direct packaging integrations.

Use an external version management tool
---------------------------------------

> An external build tool can be used that manages the version string in both the SCM and source code, either directly or via an API:

Reviewer: ```suggestion
An external build tool can be used that manages the version string in both the SCM and source code, either directly or via an API.
```

> A few tools you could use, in no particular order, and not necessarily complete:

Reviewer: Twisted has their versioning own thing too. And there's also `pbr` that I don't regard well, in comes from the OpenStack ecosystem.

`bump2version <https://pypi.org/project/bump2version>`_,

> `changes <https://pypi.org/project/changes>`_,

Reviewer 2: This project seems to be dead (no recent git commits, no release on PyPI for almost 10 years, people complaining on the tracker that it doesn't work since Python 3.7)

Author: Clearly this list needs a refresh -- anyone up to date on what's being currently maintained? In the meantime, i've removed that ones that aren't being actively maintained.

`commitizen <https://pypi.org/project/commitizen>`_,
`versioneer <https://github.com/python-versioneer/python-versioneer>`_,
`zest.releaser <https://pypi.org/project/zest.releaser>`_,

.. NOTE: are these all still being maintained??

Dedicated file for the version string
-------------------------------------

> Place the value in a simple ``VERSION`` text file and have both

Reviewer: ```suggestion
Place the value in a simple :file:`VERSION` text file and have both
```

> :file:`setup.py` and the project code read it.

Reviewer: I wonder if `version = file: VERSION` would work. Not sure if this helper works for the `version` attribute... @abravalheri do you know?

Reviewer 2: It doesn't work, you must specify in the dynamic section if you want to use that feature.

```
[tool.setuptools.dynamic]
version = {file = "VERSION"}
```

Reviewer: > It doesn't work, you must specify in the dynamic section if you want to use that feature.

Oh, I'm talking about `setup.cfg`, not `pyproject.toml`. It is processed differently.

Reviewer 2: It works in `setup.cfg`, but as previously commented the plan on setuptools side is deprecation of that file.

::

>     with open(os.path.join(mypackage_root_dir, 'VERSION')) as version_file:
>         version = version_file.read().strip()

Reviewer: The above example already uses pathlib for consistency, this should too:
```suggestion
    version = (mypackage_root_dir / 'VERSION').read_text(encoding='utf-8').strip()
```

Reviewer 2: This assumes your package is extracted to a filesystem. That's not a safe assumption.

Reviewer: Fair point. But then, `open()` still wouldn't work, right? This means the need for something like `importlib.resources.files()`.

Though, I think that it's pretty much guaranteed to be on the file system at build time. The runtime, OTOH, might need the importlib helpers.

Reviewer 2: `importlib.resources.files()` is good, but then you need the `importlib_resources` port for supporting Python < 3.9 (conditional dependency), or a compatibility fallback to stdlib [pkgutil](https://docs.python.org/3/library/pkgutil.html). To do it properly/correctly with a static file in the package it gets complicated pretty fast.

>         version = version_file.read().strip()

Reviewer 2: This method is just awful.  It totally breaks [zipapp-packaging](https://docs.python.org/3/library/zipapp.html#creating-standalone-applications-with-zipapp). Recommend removal.

Reviewer: https://github.com/pypa/packaging.python.org/pull/1273#discussion_r1317971494

Author: I think this is a overall question -- this is already under "no longer recommended" -- I left it (and others) in because I thought it would be less controversial change (or less radical anyway) -- but I'm happy to simply remove the really old and crufty ones :-)

Reviewer 2: Yes, they should be removed. The examples which import `pkg_resources` are cause deprecation warnings already.

An advantage with this technique is that it's not specific to Python.  Any
tool can read the version string.

.. warning::

>     With this approach you must make sure that the ``VERSION`` file is included in

Reviewer: ```suggestion
    With this approach you must make sure that the :file:`VERSION` file is included in
```

    all your source and binary distributions (e.g. add ``include VERSION`` to your
    :file:`MANIFEST.in`).

> No longer recommend methods

Reviewer 2: Poor english.

Reviewer: How about
```suggestion
Nearly extinct methods
```

Author: That makes the point :-)

---------------------------

> These methods rely on importing of code during the build process, or dynamically generating the version at build time. These methods are prone to errors and security issues, but you may encounter them in older code bases.

Reviewer: ```suggestion
These methods rely on importing of code during the build process and are prone to errors and security issues, but you may encounter them in older code bases.
```

I don't agree with this part of the sentence — it's perfectly fine to generate version during the build. `setuptools-scm` does just that — it makes a version by combining multiple sources — the last Git tag, distance to the last Git tag, current commit, current date, Git dirty flag, `.git_archival.txt` (for git archives and sdists).

Author: I'm not sure exactly what part you don't agree with -- but the idea is that running *custom* code at build time is not great -- which is different than using a declaration that invokes another tool that runs -- after all, *something* has to run to build a package :-) If that's not clear, then please suggest some more clear phrasing -- I'm at a loss.

Author: I've rephrased it bit -- maybe better?

Reviewer: Having reflected on this for a bit, I don't think this is right at all. Running custom code is fine, for as long as it's invoked through the standard interface. It's even documented in the official setuptools docs: https://setuptools.pypa.io/en/latest/build_meta.html#dynamic-build-dependencies-and-other-build-meta-tweaks.
Besides, it's quite mandatory to add some project-specific build code for building C-extensions. `setuptools` kinda _expects_ that `setup.py` would be used for this. But personally, I've experimented with wrappers for it on the PEP 517 level and that went well.

Hatchling also has hooks for running Python code: https://hatch.pypa.io/1.1/plugins/build-hook/#custom

           ...
        )

Generate the version string from the SCM
----------------------------------------

Most projects use a Source Code Management System (SCM), such as git or mercurial to manage the code and also manage releases.

> In order to keep the versioning of releases in sync with your package, the version number can be kept in the tags of a version control system, and automatically extracted with a tool such as

Reviewer 3: version string

`setuptools_scm <https://pypi.org/project/setuptools-scm/>`_.

.. NOTE: maybe put in an example using setuptools_scm?

.. NOTE2: Is setuptools_scm the only one now?

Use an external version management tool
---------------------------------------

An external build tool can be used that manages the version in both the SCM and source code, either directly or via an API:

A few tools you could use, in no particular order, and not necessarily complete:
`bump2version <https://pypi.org/project/bump2version>`_,
`changes <https://pypi.org/project/changes>`_,
`commitizen <https://pypi.org/project/commitizen>`_,

> `versioneer <https://github.com/python-versioneer/python-versioneer>_`

Reviewer 3: Typo. The underscore should be after the backtick.

`zest.releaser <https://pypi.org/project/zest.releaser>`_,

.. NOTE: are these all still being maintained??

Dedicated file for the version string
-------------------------------------

Place the value in a simple ``VERSION`` text file and have both
:file:`setup.py` and the project code read it.

::

    with open(os.path.join(mypackage_root_dir, 'VERSION')) as version_file:
        version = version_file.read().strip()

An advantage with this technique is that it's not specific to Python.  Any
tool can read the version.

.. warning::

    With this approach you must make sure that the ``VERSION`` file is included in
    all your source and binary distributions (e.g. add ``include VERSION`` to your
    :file:`MANIFEST.in`).

No longer recommend methods
---------------------------

> These methods rely on importing of code during the build process, or dynamically genrated the version at run time. These methods are prone to errors and security issues, but you may encounter them in older code bases.

Author: yes, `importlib{.,_}metadata` is recommended, but that's not the same  "dynamically generated the version at run time" referred to here. Better wording suggestions are welcome :-)

> These methods rely on importing of code during the build process, or dynamically genrated the version at run time. These methods are prone to errors and security issues, but you may encounter them in older code bases.

Reviewer 7: typo:

```suggestion
These methods rely on importing of code during the build process, or dynamically generated the version at run time. These methods are prone to errors and security issues, but you may encounter them in older code bases.
```

            exec(fp.read(), version)
        # later on we use: version['__version__']

.. NOTE: this example is from a (very!) old version of warehouse -- at a glance, I couldn't tell how warehouse does it now, but it's not this.

> ..    Example using this technique: `warehouse <https://github.com/pypa/warehouse/blob/64ca42e42d5613c8339b3ec5e1cb7765c6b23083/warehouse/__about__.py>`_.

Reviewer: Why comment this out?

    Example using this technique: `warehouse <https://github.com/pypa/warehouse/blob/64ca42e42d5613c8339b3ec5e1cb7765c6b23083/warehouse/__about__.py>`_.

#.  Set the value in :file:`setup.py`, and have the project code use the

>     ``importlib.metadata`` API to fetch the value at build time.

Reviewer: No, this change is incorrect. The original variant is true.

`importlib.metadata` works with an already installed dist. At build time, it's not yet installed and there's no metadata to read, vice versa — that metadata is being created so it can't be also the source. It mustn't be used during the build but in runtime (for example, via `__version__ = importlib.metadata.version(__package__)`. The original source in the described case is `setup.py`.

```suggestion
    ``importlib.metadata`` API to fetch the value at runtime.
```

    installation metadata, which is not necessarily the code that's currently
    imported.

>     If a project uses this method to fetch its version string at build time, then its

Reviewer: Same here.
```suggestion
    If a project uses this method to fetch its version string at runtime, then its
```

    installation metadata, which is not necessarily the code that's currently
    imported.

    If a project uses this method to fetch its version string at runtime, then its
    ``install_requires`` value needs to be edited to install

>     ``importlib-metadata`` on pre-3.8 versions of Python like so::

Reviewer 3: I think we can leave the detail of `importlib-metadata` pre Python 3.8 out of the guide. People can figure this out on their own. And Python 3.8 is almost EOL, right?

Reviewer 4: Well, still one year to go. Maybe we could add a todo comment here.

Author: And I think this is going to have to get updated sooner than the last update cycle anyway ...

    .. warning::

>         Although this technique was common, beware that it will fail if

Reviewer: ```suggestion
        Although this technique used to be common, beware that it will fail if
```
